fetcher: don't hold batches across boundaries

jordanlewis opened the pull request:
Updates #64906.

Previously, the cfetcher was completely blind to whether or not the keys
it retrieves from the kvfetcher lie on a batch boundary. This led to
the cfetcher retaining a pointer to 2 fetched batches at once during
fetches, since it doesn't deep copy key or value memory out of the
batches. The retained pointer is required, because the cfetcher needs to
keep the bytes of the last key and value it saw for various reasons.

Now, the kvfetcher and kvbatchfetcher participate by returning whether
or not the most recent returned key or batch lies on a *batch boundary*,
meaning that the next invocation to the fetcher would require sending a
BatchRequest to KV, getting more bytes back.

If this condition is true, the cFetcher will now deep copy its last-seen
KV instead of keeping a pointer into the most recent batch.

This improves memory usage by 2x for a brief moment, since before, the
cFetcher would always have a live pointer to both the last batch and the
next batch until it had a chance to decode the first key of the next
batch.

Commit 5c00812b5d495749133beb1e217639e03fd02246 improved this same
situation by drastically cutting down the window in which the last batch
was retained. But it wasn't quite enough. After this commit, there
should be no moment in time when the fetcher ever retains a pointer to
more than one batch at the same time.

Release note (sql change): reduce instantaneous memory usage during
scans by up to 2x.

jordanlewis committed:
colfetcher: don't hold batches across boundaries

Previously, the cfetcher was completely blind to whether or not the keys
it retrieves from the kvfetcher lie on a batch boundary. This led to
the cfetcher retaining a pointer to 2 fetched batches at once during
fetches, since it doesn't deep copy key or value memory out of the
batches. The retained pointer is required, because the cfetcher needs to
keep the bytes of the last key and value it saw for various reasons.

Now, the kvfetcher and kvbatchfetcher participate by returning whether
or not the most recent returned key or batch lies on a *batch boundary*,
meaning that the next invocation to the fetcher would require sending a
BatchRequest to KV, getting more bytes back.

If this condition is true, the cFetcher will now deep copy its last-seen
KV instead of keeping a pointer into the most recent batch.

This improves memory usage by 2x for a brief moment, since before, the
cFetcher would always have a live pointer to both the last batch and the
next batch until it had a chance to decode the first key of the next
batch.

Commit 5c00812b5d495749133beb1e217639e03fd02246 improved this same
situation by drastically cutting down the window in which the last batch
was retained. But it wasn't quite enough. After this commit, there
should be no moment in time when the fetcher ever retains a pointer to
more than one batch at the same time.

Release note (sql change): reduce instantaneous memory usage during
scans by up to 2x.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66376)
<!-- Reviewable:end -->

jordanlewis committed:
colencoding: bugfix to interleaved seek prefixes

Previously, there was a bug in the decoding methods that returned the
expected "seek prefix" for interleaved table when encountering a row
from an unwanted table. The returned seek prefix was not as precise as
it should have been: instead of returning the first key of the desired
table, it returned a key somewhere in the middle of the found table.

This did not cause any user-visible issues, since the cfetcher code is
resilient to finding an incorrect interleaved table row. But, it caused
interleaved table fetches to be less efficient than necessary.

This commit also adds explicit capacities to the returned keys and
values from the KV Fetcher, to prevent issues where callers accidentally
slice past the length of a large-capacity backing slice that backs the
returned kvs.

Release note: None

jordanlewis commented:
This PR was failing tests that uncovered an inaccuracy in the way that interleaved "seek prefixes" were previously computed. Added a first commit that fixes that problem.

jordanlewis committed:
colfetcher: use cached types array

This commit reduces the number of interface calls in the hot path for
fetching data by 1 per column per row.

Release note: None

sumeerbhola reviewed (changes requested):
> Reviewed 2 of 2 files at r1, 3 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66376)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/row/fetcher.go, line 719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAHezr9l-Dc0-tOEmq:-McAHezr9l-Dc0-tOEmr:b-hff3ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/fetcher.go#L719)):*
> ```
> gcassert:inline
> ```

Interesting. Do we run `gcassert` as part of our build?
___
*[pkg/sql/row/kv_batch_fetcher.go, line 544 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAC-kIC1CH-yxn4Bx1:-McAC-kIC1CH-yxn4Bx2:b-6n4bxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/kv_batch_fetcher.go#L544)):*
> ```Go
> 	if len(f.remainingBatches) > 0 {
> 		batchResp, f.remainingBatches = popBatch(f.remainingBatches)
> 		atBatchBoundary = len(f.remainingBatches) == 0 && !f.fetchEnd
> ```

`len(f.remainingBatches) == 0` doesn't seem to be a sufficient condition for this fetcher to have to issue a new `BatchRequest` -- there may still be responses left in `f.responses`.

The code in this file could use some more comments. It is doing something quite straightforward but it took me a while to figure out that the results from a `BatchResponse` are being extracted in two levels -- first in each response and then the `BatchResponses` in each response.
And the `origSpan` is actually staying the same for multiple `[]byte` that share the same `BatchResponses`. Hopefully the higher-layer code doesn't interpret the span as something that is fully satisfied by a single batch it has been given. Speaking of that, I don't see any write to `f.origSpan` -- where is it updated? Do we have a bug here when a scan uses `BATCH_RESPONSE` instead of `KEY_VALUES`?
___
*[pkg/sql/row/kv_batch_fetcher.go, line 559 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAEuE20MPZyeJOVR9C:-McAEuE20MPZyeJOVR9D:b-47tw5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/kv_batch_fetcher.go#L559)):*
> ```Go
> 				batchResp, f.remainingBatches = popBatch(t.BatchResponses)
> 			}
> 			atBatchBoundary = len(f.responses) == 0 && len(f.remainingBatches) == 0 && !f.fetchEnd
> ```

how about a separate function with this condition -- I am assuming it will get inlined.
___
*[pkg/sql/row/kv_fetcher.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAGPlh6sRsk_8KMUgv:-McAGPlh6sRsk_8KMUgw:bpvlr9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/kv_fetcher.go#L105)):*
> ```Go
> 		f.newSpan = false
> 		nKvs := len(f.kvs)
> 		if nKvs != 0 {
> ```

Can you add a comment that f.kvs is not used when f.batchResponse is non-nil and vice versa.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice find! I have some nits, but I'll defer to Sumeer for approval.

> Reviewed 2 of 2 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-:-McAguqw7uKJ6poSti0z:b-c6pl0i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @sumeerbhola)

___
*[pkg/sql/row/fetcher.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAghnpD5dEQxjwCFrA:-McAghnpD5dEQxjwCFrB:b-5lr8jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/fetcher.go#L55)):*
> ```Go
> 	// version - both must be handled by calling code.
> 	nextBatch(ctx context.Context) (ok bool, kvs []roachpb.KeyValue,
> 		batchResponse []byte, origSpan roachpb.Span, moreKeys bool, err error)
> ```

nit: `s/moreKeys/atBatchBoundary/`, right?
___
*[pkg/sql/row/fetcher.go, line 719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAHezr9l-Dc0-tOEmq:-McAel2V8u4SsG_pfWN_:b-8v5lvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/fetcher.go#L719)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> ```
> gcassert:inline
> ```

Interesting. Do we run `gcassert` as part of our build?
</blockquote></details>

It is `TestGCAssert` linter.

Jordan, you'll need to include `sql/row` package to run the GCAssert against. Or maybe we should switch to opt-in by default for the linter? Will it be too slow?
___
*[pkg/sql/row/fetcher.go, line 722 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAfysL6Qe65WpHlP1M:-McAfysL6Qe65WpHlP1N:b-4hvvqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/fetcher.go#L722)):*
> ```Go
> // gcassert:inline
> func (rf *Fetcher) setNextKV(kv roachpb.KeyValue, needsCopy bool) {
> 	if needsCopy {
> ```

super nit: I like having symmetry if possible, maybe refactor one of `setNextKV` methods to have the structure of the other (e.g. having `if !needsCopy` check here and return early).
___
*[pkg/sql/row/kv_batch_fetcher.go, line 559 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAEuE20MPZyeJOVR9C:-McAgW510Rf8AT_aPWiF:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/kv_batch_fetcher.go#L559)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

how about a separate function with this condition -- I am assuming it will get inlined.
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Alright, PTAL. Sorry for the long delay, thanks for the reviews. I think I was able to simplify this by a decent amount.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-:-MdczAJf0EbYA5DvYtTd:bx4hvtp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @yuzefovich)

___
*[pkg/sql/row/fetcher.go, line 719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAHezr9l-Dc0-tOEmq:-Mdci82c12GvrHF68OJs:b-rcsod) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/fetcher.go#L719)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It is `TestGCAssert` linter.

Jordan, you'll need to include `sql/row` package to run the GCAssert against. Or maybe we should switch to opt-in by default for the linter? Will it be too slow?
</blockquote></details>

Done.

Yeah I think it would be too slow to turn on for all packages.
___
*[pkg/sql/row/fetcher.go, line 722 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAfysL6Qe65WpHlP1M:-MdcimI32wOg51w2LfDn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/fetcher.go#L722)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

super nit: I like having symmetry if possible, maybe refactor one of `setNextKV` methods to have the structure of the other (e.g. having `if !needsCopy` check here and return early).
</blockquote></details>

Done.
___
*[pkg/sql/row/kv_batch_fetcher.go, line 544 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAC-kIC1CH-yxn4Bx1:-MdclCgCEjILgN0GeXTB:b-yns65t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/kv_batch_fetcher.go#L544)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

`len(f.remainingBatches) == 0` doesn't seem to be a sufficient condition for this fetcher to have to issue a new `BatchRequest` -- there may still be responses left in `f.responses`.

The code in this file could use some more comments. It is doing something quite straightforward but it took me a while to figure out that the results from a `BatchResponse` are being extracted in two levels -- first in each response and then the `BatchResponses` in each response.
And the `origSpan` is actually staying the same for multiple `[]byte` that share the same `BatchResponses`. Hopefully the higher-layer code doesn't interpret the span as something that is fully satisfied by a single batch it has been given. Speaking of that, I don't see any write to `f.origSpan` -- where is it updated? Do we have a bug here when a scan uses `BATCH_RESPONSE` instead of `KEY_VALUES`?
</blockquote></details>

Thinking this through some more, I actually always want this function to be returning true, so I've removed the parameter. The purpose of this exercise was to allow higher-level code to know, given a returned `roachpb.KeyValue`, whether it was the final reference to a larger backing byte slice - indicating that the next call to `NextKV` would potentially ask for another larger backing byte slice before the caller had a chance to clear its references to the previous larger backing byte slice.

This is always true for every byte slice batch - since for every batch, the only remaining pointers to that batch once its finished is a returned KV. So we can handle this logic completely in the layer that slices up BatchResponses into KVs one by one (the kv_fetcher.go code).

I added some comments about what is actually going on in this loop. I agree it's confusing, the recursion doesn't help although I'm not sure the best way to structure the loop without the recursive call at the end.

You're right about f.origSpan: we don't seem to update it properly. It actually turns out that nobody uses this return value - I deleted it altogether because it's confusing and broken to your point.
___
*[pkg/sql/row/kv_batch_fetcher.go, line 559 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAEuE20MPZyeJOVR9C:-Mdcy7oz6VmnP8_N4hI3:b-zao5au) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/kv_batch_fetcher.go#L559)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

+1
</blockquote></details>

I was able to delete this altogether.
___
*[pkg/sql/row/kv_fetcher.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-McAGPlh6sRsk_8KMUgv:-MdcyAKW8T4J2aXrxtGq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae5b21879672b1fccf1bf4dd19dfc768d18e4315/pkg/sql/row/kv_fetcher.go#L105)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add a comment that f.kvs is not used when f.batchResponse is non-nil and vice versa.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 5 files at r3, 3 of 6 files at r6, 3 of 3 files at r8, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66376#-:-MeArvwrCf4ZWO5utwSk:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola and @yuzefovich)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTR!

bors r+

jordanlewis commented:
(this is just a test, not really backporting)

blathers backport 21.1
